fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/AbstractArrayConstructorValue.java

```diff
@@ -0,0 +1,229 @@
+        public <M extends Message> Object eval(@Nonnull final FDBRecordStoreBase<M> store, @Nonnull final EvaluationContext context, @Nullable final FDBRecord<M> record, @Nullable final M message) {
+            return StreamSupport.stream(getChildren().spliterator(), false)
+                    .map(child -> child.eval(store, context, record, message))
+                    .collect(ImmutableList.toImmutableList());
+        }
+
+        @Nonnull
+        @Override
+        public LightArrayConstructorValue withChildren(final Iterable<? extends Value> newChildren) {
+            Verify.verify(!Iterables.isEmpty(newChildren));
+            Verify.verifyNotNull(resolveElementType(newChildren).equals(getElementType()));
+            return new LightArrayConstructorValue(ImmutableList.copyOf(newChildren), getElementType());
+        }
+    }
+
+    @SuppressWarnings("java:S2160")
+    public static class ArrayConstructorValue extends AbstractArrayConstructorValue {
+
+        public ArrayConstructorValue(@Nonnull final List<? extends Value> children, @Nonnull final Type elementType) {
+            super(children, elementType);
+        }
+
+        @Nullable
+        @Override
+        @SuppressWarnings("java:S6213")
+        public <M extends Message> Object eval(@Nonnull final FDBRecordStoreBase<M> store, @Nonnull final EvaluationContext context, @Nullable final FDBRecord<M> record, @Nullable final M message) {
+            final DynamicMessage.Builder resultMessageBuilder = newMessageBuilderForType(context.getDynamicSchema());
+            final Descriptors.Descriptor descriptorForType = resultMessageBuilder.getDescriptorForType();
+
+            return StreamSupport.stream(getChildren().spliterator(), false)
+                    .map(child -> {
+                        final Object childResultElement = child.eval(store, context, record, message);
+                        final DynamicMessage.Builder helperMessageBuilder = DynamicMessage.newBuilder(descriptorForType);
+                        if (childResultElement != null) {
+                            helperMessageBuilder.setField(descriptorForType.findFieldByNumber(1), childResultElement);
+                        }
+                        return helperMessageBuilder.build();
+                    })
+                    .collect(ImmutableList.toImmutableList());
+        }
+
+        @Nonnull
+        @Override
+        public ArrayConstructorValue withChildren(final Iterable<? extends Value> newChildren) {
+            Verify.verify(!Iterables.isEmpty(newChildren));
+            Verify.verifyNotNull(resolveElementType(newChildren).equals(getElementType()));
+            return new ArrayConstructorValue(ImmutableList.copyOf(newChildren), getElementType());
+        }
+
+        @Nonnull
+        private DynamicMessage.Builder newMessageBuilderForType(@Nonnull DynamicSchema dynamicSchema) {
+            return dynamicSchema.newMessageBuilder(dynamicSchema.getProtoTypeName(getResultType()));
```

Comment:
It seems as though the _names_ of the generated types are not a first-class part of the contract. So I think using `newMessageBuilder(Type)` would be preferred.

Reply:
done.